connect/mirror/src/test/java/org/apache/kafka/connect/mirror/MirrorConnectorsIntegrationTest.java

```diff
@@ -150,23 +152,73 @@ public void setup() throws IOException {
         mm2Props.put("backup.bootstrap.servers", backup.kafka().bootstrapServers());
         mm2Config = new MirrorMakerConfig(mm2Props);
 
+        // we wait for the connector and tasks to come up for each connector, so that when we do the
+        // actual testing, we are certain that the tasks are up and running; this will prevent
+        // flaky tests where the connector and tasks didn't start up in time for the tests to be
+        // run
         backup.configureConnector("MirrorSourceConnector", mm2Config.connectorBaseConfig(new SourceAndTarget("primary", "backup"),
             MirrorSourceConnector.class));
+        waitForConnectorToBeUp(backup, "MirrorSourceConnector");
+        waitForTasksToBeUp(backup, "MirrorSourceConnector");
 
         backup.configureConnector("MirrorCheckpointConnector", mm2Config.connectorBaseConfig(new SourceAndTarget("primary", "backup"),
             MirrorCheckpointConnector.class));
+        waitForConnectorToBeUp(backup, "MirrorCheckpointConnector");
+        waitForTasksToBeUp(backup, "MirrorCheckpointConnector");
 
         backup.configureConnector("MirrorHeartbeatConnector", mm2Config.connectorBaseConfig(new SourceAndTarget("primary", "backup"),
             MirrorHeartbeatConnector.class));
+        waitForConnectorToBeUp(backup, "MirrorHeartbeatConnector");
+        waitForTasksToBeUp(backup, "MirrorHeartbeatConnector");
 
         primary.configureConnector("MirrorSourceConnector", mm2Config.connectorBaseConfig(new SourceAndTarget("backup", "primary"),
             MirrorSourceConnector.class));
+        waitForConnectorToBeUp(primary, "MirrorSourceConnector");
+        waitForTasksToBeUp(primary, "MirrorSourceConnector");
 
         primary.configureConnector("MirrorCheckpointConnector", mm2Config.connectorBaseConfig(new SourceAndTarget("backup", "primary"),
             MirrorCheckpointConnector.class));
+        waitForConnectorToBeUp(primary, "MirrorCheckpointConnector");
+        waitForTasksToBeUp(primary, "MirrorCheckpointConnector");
 
         primary.configureConnector("MirrorHeartbeatConnector", mm2Config.connectorBaseConfig(new SourceAndTarget("backup", "primary"),
             MirrorHeartbeatConnector.class));
+        waitForConnectorToBeUp(primary, "MirrorHeartbeatConnector");
+        waitForTasksToBeUp(primary, "MirrorHeartbeatConnector");
+
+    }
+
+    private void waitForConnectorToBeUp(EmbeddedConnectCluster connectCluster, String connName)
+        throws InterruptedException {
+        Collection<String> connectors = connectCluster.connectors();
+        while (!connectors.contains(connName)) {
+            // this means the connector hasn't been registered yet in the REST endpoint, so we
+            // sleep for 3 seconds and try again
+            Thread.sleep(3_000);
+            connectors = connectCluster.connectors();
+        }
+    }
+
+    private void waitForTasksToBeUp(EmbeddedConnectCluster connectCluster, String connectorName)
```

Comment:
nit: `waitForTasksToBeUp` -> `waitForTasksToStart`?

Reply:
That does make it a lot more readable, will make the change.

Reply:
I'd suggest copying and using what we've been using elsewhere to assert that a connector and its tasks are up and running: 

https://github.com/apache/kafka/blob/trunk/connect/runtime/src/test/java/org/apache/kafka/connect/integration/RebalanceSourceConnectorsIntegrationTest.java#L324-L337

It works find with the `waitForCondition` method that accepts a timeout. 
Eventually (maybe soon) this type of assertion will go to `EmbeddedConnectCluster` and will be made available for every Connect integration test. However, I suggest we don't take on this refactoring now and we just copy the method from the test above.

Reply:
Thanks for the suggestion @kkonstantine - this is super useful! I'll incorporate it into the code.